metadata-ingestion/src/datahub/utilities/sqlglot_lineage.py

```diff
@@ -0,0 +1,754 @@
+    table_schemas_resolved: int
+
+    column_error: Optional[Exception]
+
+
+class SqlParsingResult(BaseModel):
+    query_type: QueryType = QueryType.UNKNOWN
+
+    in_tables: List[Urn]
+    out_tables: List[Urn]
+
+    column_lineage: Optional[List[ColumnLineageInfo]]
+
+    # TODO include formatted original sql logic
+    # TODO include list of referenced columns
+
+    debug_info: SqlParsingDebugInfo = pydantic.Field(
+        default_factory=lambda: SqlParsingDebugInfo(
+            confidence=0,
+            tables_discovered=0,
+            table_schemas_resolved=0,
+            column_error=None,
+        ),
+        exclude=True,
+    )
+
+
+def _parse_statement(sql: str, dialect: str) -> sqlglot.Expression:
+    statement = sqlglot.parse_one(
+        sql, read=dialect, error_level=sqlglot.ErrorLevel.RAISE
+    )
+    return statement
+
+
+def _table_level_lineage(
+    statement: sqlglot.Expression,
+    dialect: str,
+) -> Tuple[Set[_TableName], Set[_TableName]]:
+    def _raw_table_name(table: sqlglot.exp.Table) -> _TableName:
+        return _TableName.from_sqlglot_table(table, dialect=dialect)
+
+    # Generate table-level lineage.
+    modified = {
+        _raw_table_name(expr.this)
+        for expr in statement.find_all(
+            sqlglot.exp.Create,
+            sqlglot.exp.Insert,
+            sqlglot.exp.Update,
+            sqlglot.exp.Delete,
+            sqlglot.exp.Merge,
+        )
+        # In some cases like "MERGE ... then INSERT (col1, col2) VALUES (col1, col2)",
+        # the `this` on the INSERT part isn't a table.
+        if isinstance(expr.this, sqlglot.exp.Table)
+    }
+
+    tables = (
+        {_raw_table_name(table) for table in statement.find_all(sqlglot.exp.Table)}
+        # ignore references created in this query
+        - modified
+        # ignore CTEs created in this statement
+        - {
+            _TableName(database=None, schema=None, table=cte.alias_or_name)
+            for cte in statement.find_all(sqlglot.exp.CTE)
+        }
+    )
+    # TODO: If a CTAS has "LIMIT 0", it's not really lineage, just copying the schema.
+
+    return tables, modified
+
+
+class SchemaResolver:
+    def __init__(
+        self,
+        *,
+        platform: str,
+        platform_instance: Optional[str] = None,
+        env: str = DEFAULT_ENV,
+        graph: Optional[DataHubGraph] = None,
+        _cache_filename: Optional[pathlib.Path] = None,
+    ):
+        # TODO handle platforms when prefixed with urn:li:dataPlatform:
+        self.platform = platform
+        self.platform_instance = platform_instance
+        self.env = env
+
+        self.graph = graph
+
+        # Init cache, potentially restoring from a previous run.
+        shared_conn = None
+        if _cache_filename:
+            shared_conn = ConnectionWrapper(filename=_cache_filename)
+        self._schema_cache: FileBackedDict[Optional[SchemaInfo]] = FileBackedDict(
+            shared_connection=shared_conn,
+        )
```

Comment:
Doesn't look like we ever close this, are we just relying on garbage collection for that? I could see that taking a while because references may be kept in memory by the `lru_cache`

Reply:
fixed in the follow up by explicitly calling self._make_schema_resolver.cache_clear()